sdk/nodejs/core/zMixins.ts

```diff
@@ -0,0 +1,37 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { ServiceClientCredentials } from "@azure/ms-rest-js";
+import * as msnodeauth from "@azure/ms-rest-nodeauth";
+import * as config from "../config";
+
+/**
+ * Obtain credentials to query Azure Management API. Depending on the environment configuration, this
+ * are either based on MSI, a service principal, or Azure CLI user credentials.
+ */
+export async function getServiceClientCredentials(): Promise<ServiceClientCredentials> {
+    let credentials: ServiceClientCredentials;
+
+    if (config.useMsi) {
+        credentials = await msnodeauth.loginWithAppServiceMSI({ msiEndpoint: config.msiEndpoint });
+    } else if (config.clientId && config.clientSecret && config.tenantId) {
+        credentials = await msnodeauth.loginWithServicePrincipalSecret(
+            config.clientId, config.clientSecret, config.tenantId);
+    } else {
+        // `create()` will throw an error if the Az CLI is not installed or `az login` has never been run.
+        credentials = await msnodeauth.AzureCliCredentials.create();
+    }
```

Comment:
Are we sure these are the complete set of configuration options supported by the Azure provider?  I'm a little nervous re-creating this logic outside of the provider.  Could we perhaps cross-reference to the primary source for this logic? (Ideally in the backing provider to make sure it's really the same as what the user is seeing for other `@pulumi/azure` operations.

Reply:
Yeah, me too... I don't know of the source of truth for this.
Ideally, I'd use some function provided by Pulumi core/Terraform which would give me an access token that the provider is already using. Is there something like that? I asked on Slack once but didn't get an answer.